crates/optimism/primitives/src/op_tx_type.rs

```diff
@@ -0,0 +1,190 @@
+//! newtype pattern on op_alloy_consensus::OpTxType.
+//! OpTxType that implements reth_primitives_traits::TxType.
+//! This type is required because a Compact impl is needed on the deposit tx type.
+
+use alloy_primitives::{U64, U8};
+use alloy_rlp::{Decodable, Encodable, Error};
+use bytes::BufMut;
+use core::fmt::Debug;
+use derive_more::{
+    derive::{From, Into},
+    Display,
+};
+use op_alloy_consensus::OpTxType as AlloyOpTxType;
+use std::convert::TryFrom;
+
+/// Wrapper type for `AlloyOpTxType` to implement `TxType` trait.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Display, Ord, Hash, From, Into)]
+#[from(AlloyOpTxType)]
+#[into(u8)]
+pub struct OpTxType(AlloyOpTxType);
+
+impl From<OpTxType> for U8 {
+    fn from(tx_type: OpTxType) -> Self {
+        U8::from(u8::from(tx_type))
+    }
+}
+
+impl TryFrom<u8> for OpTxType {
+    type Error = Error;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        AlloyOpTxType::try_from(value)
+            .map(OpTxType)
+            .map_err(|_| Error::Custom("Invalid transaction type"))
+    }
+}
```

Comment:
this should be possible to derive with `derive_more::TryFrom`, since `TryFrom<u8>` is impl for `OpTxType` already

Reply:
yes but the problem is the original `OpTxType` is enum, and `TryFrom` is defined for enums. Our `OpTxType` is struct, and i don't know a way to make it work for structs

Reply:
ah yes, sorry, we had this